src/hotspot/cpu/riscv/riscv.ad

```diff
@@ -7941,113 +7947,154 @@ instruct xorL_reg_imm(iRegLNoSp dst, iRegL src1, immLAdd src2) %{
 // ============================================================================
 // MemBar Instruction
 
-instruct load_fence() %{
+// ZTSO
+
+instruct no_membar_rvtso() %{
+  predicate(UseZtso);
   match(LoadFence);
-  ins_cost(ALU_COST);
+  match(StoreFence);
+  match(StoreStoreFence);
+  match(MemBarAcquire);
+  match(MemBarRelease);
+  match(MemBarStoreStore);
+  match(MemBarAcquireLock);
+  match(MemBarReleaseLock);
 
-  format %{ "#@load_fence" %}
+  ins_cost(0);
+
+  size(0);
 
+  format %{ "no_membar_rvtso elided/tso (empty encoding)" %}
   ins_encode %{
-    __ membar(MacroAssembler::LoadLoad | MacroAssembler::LoadStore);
+    __ block_comment("no_membar_rvtso");
   %}
-  ins_pipe(pipe_serial);
+  ins_pipe(real_empty);
 %}
 
-instruct membar_acquire() %{
-  match(MemBarAcquire);
-  ins_cost(ALU_COST);
+instruct membar_volatile_rvtso() %{
+  predicate(UseZtso);
+  match(MemBarVolatile);
+  ins_cost(VOLATILE_REF_COST);
 
-  format %{ "#@membar_acquire\n\t"
-            "fence ir iorw" %}
+  format %{ "membar_volatile_rvtso\n\t"
+            "fence w, r"%}
 
   ins_encode %{
-    __ block_comment("membar_acquire");
-    __ membar(MacroAssembler::LoadLoad | MacroAssembler::LoadStore);
+    __ block_comment("membar_volatile_rvtso");
+    __ membar(MacroAssembler::StoreLoad);
   %}
 
-  ins_pipe(pipe_serial);
+  ins_pipe(pipe_slow);
 %}
 
-instruct membar_acquire_lock() %{
-  match(MemBarAcquireLock);
+instruct unnecessary_membar_volatile_rvtso() %{
+  predicate(UseZtso && Matcher::post_store_load_barrier(n));
+  match(MemBarVolatile);
   ins_cost(0);
 
-  format %{ "#@membar_acquire_lock (elided)" %}
-
+  size(0);
+  
+  format %{ "unnecessary_membar_volatile_rvtso (unnecessary so empty encoding)" %}
   ins_encode %{
-    __ block_comment("membar_acquire_lock (elided)");
+    assert(UseZtso, "Must be");
+    __ block_comment("unnecessary_membar_volatile_rvtso");
   %}
-
-  ins_pipe(pipe_serial);
+  ins_pipe(real_empty);
 %}
 
-instruct store_fence() %{
-  match(StoreFence);
-  ins_cost(ALU_COST);
+// RVWMO
+
+instruct membar_rvwmo_aqcuire() %{
```

Comment:
Maybe it's better to put `_rvwmo` as a suffix? Like `membar_aqcuire_rvwmo`. Similar for other match rules like `membar_rvwmo_release`, `membar_rvwmo_storestore`, `membar_rvwmo_lock` and `membar_rvwmo_volatile`.

Reply:
Fixed